### PR TITLE
consul: add support for service weight

### PR DESCRIPTION
**Pull request by mvegter**
## Manual verification
```
No `weights`
```
```
$ curl -s http://127.0.0.1:8500/v1/agent/services  | jq '.[] | select(.Service=="global-redis-check") | .Weights'
{
  "Passing": 1,
  "Warning": 1
}

$ dig @127.0.0.1 -p 8600 global-redis-check.service.consul SRV +short
1 1 27982 7f000001.addr.dc1.consul.
```
---
```
weights {}
```
```
$ ./bin/nomad job plan ./job.hcl
+/- Job: "redis"
+/- Task Group: "cache" (1 in-place update)
  +/- Task: "redis" (forces in-place update)
    +/- Service {
        Address:           ""
        ...
      + Weights {
        + Passing: "1"
        + Warning: "1"
        }
        }
```
```
$ curl -s http://127.0.0.1:8500/v1/agent/services  | jq '.[] | select(.Service=="global-redis-check") | .Weights'
{
  "Passing": 1,
  "Warning": 1
}	
```
---
```
weights {
  passing = 5
}
```
```
$ ./bin/nomad job plan ./job.hcl
+/- Job: "redis"
+/- Task Group: "cache" (1 in-place update)
  +/- Task: "redis" (forces in-place update)
    +/- Service {
          Address:           ""
          ...
      +/- Weights {
        +/- Passing: "1" => "5"
            Warning: "1"
          }
        }
```
```
$ curl -s http://127.0.0.1:8500/v1/agent/services  | jq '.[] | select(.Service=="global-redis-check") | .Weights'
{
  "Passing": 5,
  "Warning": 1
}

$ dig @127.0.0.1 -p 8600 global-redis-check.service.consul SRV +short
1 5 30957 7f000001.addr.dc1.consul.
```

### CODE REVIEW

**Comment by mvegter** [comment]
```
=== NAME  TestVaultCompat/testVaultVersions
    vaultcompat_test.go:326: download: installing vault at /tmp/vault-bins/1.18.0
    vaultcompat_test.go:332: 
        vaultcompat_test.go:332: expected nil error
        ↪ error: exit status 1
        ↪ PostScript | annotation ↷
        	failed to download vault 1.18.0: hc-install: will install vault@1.18.0
        failed to install vault@1.18.0: unexpected Content-Type: "application/vnd+hashicorp.releases-api.v1+json"
```
Seems like CI is flaky here, unrelated to the changeset

**Comment by tgross** [comment]
Hi @mvegter! Just wanted to let you know we do see this PR and appreciate that you're trying to get tests green. We're a little swamped with some post-1.9.0 mop-up but we'll review this as soon as feasible. Thanks!

**Comment by mvegter** [comment]
hey @tgross , thanks for the heads up. After some more testing I noticed a small issue where the `agentServiceUpdateRequired` returned false when going from having the `weights` stanza to unsetting it. It should now be fixed and added some more tests to cover this as well, I see the last failing pipelines turning green as we speak. Looking forward to your feedback !

**Comment by Juanadelacuesta** [comment]
Hello @mvegter thank you for putting more work into your PR, @philrenaud can give you a hand on the Vercel problems.

File: api/tasks.go
Comment (tgross) [inline api/tasks.go:1273]: This type should probably be defined in `api/services.go` instead of here.

File: command/agent/consul/connect.go
Comment (tgross) [inline command/agent/consul/connect.go:108]: This function is only used in one place and it's being defined in an unrelated file. Let's move this to where it's being used in `service_client.go`.

File: command/agent/job_endpoint.go
Comment (tgross) [inline command/agent/job_endpoint.go:1755]: The function has a nil check itself so I think we can skip this conditional

```suggestion		
		out[i].Weights = apiWorkloadWeightsToStructs(s.Weights)
```

File: nomad/structs/services.go
Comment (tgross) [inline nomad/structs/services.go:682]: The `Copy` has a nil check, so we can safely skip this conditional.

```suggestion	
	ns.Weights = s.Weights.Copy()
```

File: website/content/docs/job-specification/service.mdx
Comment (tgross) [inline website/content/docs/job-specification/service.mdx:103]: It's good to have a link to the Consul docs but to allow for drift (and to account for the subtle difference between block and map syntax) let's inline the field definitions:

```suggestion
- `weights` <code>(Weights: nil)</code> - Specifies how a service instance is
  weighted in a DNS SRV request based on the service's health status, as
  described in the Consul [weights][] documentation. Only available where
  `provider = "consul"` The `weight` block supports the following fields:
  - `passing` <code>int: 1</code> - The weight of services in passing state.
  - `warning` <code>int: 1</code> - The weight of services in warning state.

```

File: nomad/structs/services.go
Comment (tgross) [inline nomad/structs/services.go:1151]: If we call this struct `ConsulWeights`, we're stuck with it if we later decide to support DNS and weights in Nomad service discovery. Let's call this `ServiceWeights` instead (that'll ripple through the rest of the PR as well).

File: command/agent/consul/unit_test.go
Comment (tgross) [inline command/agent/consul/unit_test.go:49]: where does this change to the test workload get exercised?

**Review by tgross — CHANGES_REQUESTED** [review]
Hi @mvegter! This is looking really good. I've left some comments but once those are resolved I think we'll be ready to merge.

**Review by tgross — APPROVED** [review]
LGTM! Once CI has finished we'll get this merged and it'll ship in the next release of Nomad. Thanks @mvegter!

**Comment by mvegter** [comment]
Thanks @tgross , looking forward to the next release of 1.9.x !

**Pull request merged**
